### PR TITLE
Double Round Robin

### LINKED ISSUE #87: Double Round Robin (DRR)

### PR DESCRIPTION
**Pull request by ketphan02**
Sample output:
```
Class size: 8
Team 0: [1, 3, 0], Project: 1
Team 1: [6, 7, 4], Project: 2
Team 2: [2, 5], Project: 3
```

closes #87

### CODE REVIEW

File: api/ai/new/double_round_robin_algorithm/double_round_robin_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            selected_students,
+        )
+        self._round_robin(
+            positive_utilities,
+            len(positive_utility_students),
+            allocation,
+            selected_students,
+        )
+
+        team_list: List[Team] = []
```
Comment (Opeyem1a) [inline api/ai/new/double_round_robin_algorithm/double_round_robin_algorithm.py:115]: try running this and right before this just `print(self.teams)` to see what happens, I'm curious about something

**Review by Opeyem1a — COMMENTED** [review]
Looks fine but see the comment I made - if I remember correctly, the team names are created before being sent off to the algorithm, so there would already be a `self.teams` with empty teams that you can just populate

File: api/ai/new/double_round_robin_algorithm/double_round_robin_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        self._round_robin(
+            positive_utilities,
+            len(positive_utility_students),
+            allocation,
+            selected_students,
+        )
+
+        for team_idx, (project_id, students) in enumerate(allocation.items()):
+            self.teams[team_idx].students = students
+            self.teams[team_idx].project_id = project_id
```
Comment (Opeyem1a) [inline api/ai/new/double_round_robin_algorithm/double_round_robin_algorithm.py:119]: it's merge-able but now that I look at it, you're using `projects` for their requirements and their ids, but `Team` can have both? Why not have the whole algorithm just be using teams instead?

**Review by Opeyem1a — APPROVED** [review]

File: api/ai/new/double_round_robin_algorithm/double_round_robin_algorithm.py (hunk 2 above)
Comment (ketphan02) [reply api/ai/new/double_round_robin_algorithm/double_round_robin_algorithm.py:119]: You know what, you're not wrong! Implementing 👀

File: api/ai/new/double_round_robin_algorithm/double_round_robin_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+
+class DoubleRoundRobinAlgorithm(Algorithm):
+    def __init__(
+        self,
+        algorithm_options: DoubleRoundRobinAlgorithmOptions,
+        team_generation_options: TeamGenerationOptions,
+    ):
+        super().__init__(algorithm_options, team_generation_options)
+
```
Comment (ketphan02) [inline api/ai/new/double_round_robin_algorithm/double_round_robin_algorithm.py:48]: should validate `team_generation_options`

**Comment by ketphan02** [comment]
Dependency: #168

File: api/ai/new/double_round_robin_algorithm/utils.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+from typing import List, Dict
+
+from api.ai.new.double_round_robin_algorithm.custom_models import Utility
+from api.models.enums import RequirementOperator
+from api.models.project import ProjectRequirement
+from api.models.student import Student
+from api.models.team import Team
+
+
+def requirement_met_by_student(requirement: ProjectRequirement, student: Student):
```
Comment (Opeyem1a) [inline api/ai/new/double_round_robin_algorithm/utils.py:10]: I''m not sure why we need this?

**Review by Opeyem1a — CHANGES_REQUESTED** [review]
a question

**Pull request merged**